File: ruby/roman_to_integer.rb
MAP = {
    'I' => 1,
    'V' => 5,
    'X' => 10,
    'L' => 50,
    'C' => 100,
    'D' => 500,
    'M' => 1000
}

# @param {String} s
# @return {Integer}
def roman_to_int(s)
    res = 0
    i = 0
    while i < s.length
        if i == s.length - 1
            res += MAP[s[i]]
            i += 1
        else
            x = MAP[s[i]]
            y = MAP[s[i + 1]]
            res += x < y ? y - x : x
            i += x < y ? 2 : 1
        end
    end
    res
end

p roman_to_int('I')
p roman_to_int('IV')
p roman_to_int('X')

File: java/RomanToInteger.java
/**
 * Author: Chih-Jye Wang
 * Date  : Dec 17, 2013
 *
 * Given a roman numeral, convert it to an integer.
 *
 * Input is guaranteed to be within the range from 1 to 3999.
 */
public class RomanToInteger
{
    public int romanToInt(String s)
    {
        int i   = 0;
        int sum = 0;
        while(i < s.length())
        {
            //Last character of the string
            if(i == s.length() - 1)
                return sum + digitValue(s.charAt(i));
            
            //Get the value of current and next positions
            int current = digitValue(s.charAt(i));
            int next    = digitValue(s.charAt(i + 1));
            //If current is less than next, subtractive notation;
            //Else normal additive notation.
            if(current < next)
            {    
                sum += next - current;
                i   += 2;
            }    
            else
            {
                sum += current;
                i++;
            }        
        }
        return sum; 
    }
    
    int digitValue(char roman)
    {
        switch(roman)
        {
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
        }
        return 0;
    }
}